tests/plenary/async_lib/semaphore_spec.lua

```diff
@@ -12,6 +12,40 @@ a.describe('semaphore', function()
     eq(status, false)
   end)
 
-  a.it('should count properly', function()
+  a.it('should acquire a permit if available', function ()
+    local sem = a.util.Semaphore.new(1)
+    local permit = await(sem:acquire())
+    assert(permit ~= nil)
+  end)
+
+  a.it('should block if no permit is available', function ()
+    local sem = a.util.Semaphore.new(1)
+    await(sem:acquire())
+    
+    local completed = false
+    local blocking = async(function ()
+      await(sem:acquire())
+      completed = true
+    end)
+    a.run(blocking())
+
+    eq(completed, false)
+  end)
+
+  a.it('should permit the next waiting client when a permit is released', function () 
+    local sem = a.util.Semaphore.new(1)
+    local permit = await(sem:acquire())
+
+    local completed = false
+    local blocking = async(function ()
+      await(sem:acquire())
+      completed = true
+    end)
+
+    local blocking_future = blocking()
+    permit:forget()
+    a.util.block_on(blocking_future)
```

Comment:
why not `a.run`?

Reply:
gap in knowledge :D can I be sure that the future is run to completion with `a.run`? blocking just seemed like the intuitive thing to do.

Reply:
actually my bad `await` would actually be the right thing to use here. You are right, `a.run` will not necessarily run it to completion within the function but it probably will in this case because we are not doing any real async such as io, we are just using coroutine abstractions. `await` will always run to completion. The test is automatically blocked on also.